dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,429 @@
+        } else {
+            S thenStore = input.getThenStore();
+            sbStore.append("then=");
+            sbStore.append(visualizeStore(thenStore));
+            S elseStore = input.getElseStore();
+            sbStore.append(", else=");
+            sbStore.append(visualizeStore(elseStore));
+        }
+        sbStore.append(escapeCharacter).append("~~~~~~~~~").append(escapeCharacter);
+        return sbStore.toString();
+    }
+
+    /**
+     * Helper method to simplify visualizing a special Block; it is useful when implementing
+     * a custom CFGVisualizer.
+     *
+     * @param sbb The special {@link Block}.
+     * @param separator The separator String to use.
+     * @return The String representation of the special {@link Block}.
+     */
+    protected String visualizeSpecialBlockHelper(SpecialBlock sbb, String separator) {
+        String specialBlock = "";
+        switch (sbb.getSpecialType()) {
+            case ENTRY:
+                specialBlock = "<entry>" + separator;
+                break;
+            case EXIT:
+                specialBlock = "<exit>" + separator;
+                break;
+            case EXCEPTIONAL_EXIT:
+                specialBlock = "<exceptional-exit>" + separator;
+                break;
+        }
+        return specialBlock;
+    }
+
+    /**
+     * Helper method called by {@link #visualizeBlockHelper}. If possible, get the last
+     * {@link Node} of a {@link Block}.
+     *
+     * @param bb The {@link Block}.
+     * @return The last {@link Node} or {@code null}.
+     */
+    protected Node getLastNode(Block bb) {
+        Node lastNode;
+        switch (bb.getType()) {
+            case REGULAR_BLOCK:
+                List<Node> blockContents = ((RegularBlock) bb).getContents();
+                lastNode = blockContents.get(blockContents.size() - 1);
+                break;
+            case EXCEPTION_BLOCK:
+                lastNode = ((ExceptionBlock) bb).getNode();
+                break;
+            default:
+                lastNode = null;
+        }
+        return lastNode;
+    }
+
+    /**
+     * Generate the order of processing {@link Block}s.
+     *
+     * @param cfg The current control flow graph.
+     * @return The IdentityHashMap which maps from {@link Block}s to their orders.
+     */
+    protected IdentityHashMap<Block, List<Integer>> getProcessOrder(ControlFlowGraph cfg) {
+        IdentityHashMap<Block, List<Integer>> depthFirstOrder = new IdentityHashMap<>();
+        int count = 1;
+        for (Block b : cfg.getDepthFirstOrderedBlocks()) {
+            depthFirstOrder.computeIfAbsent(b, k -> new ArrayList<>());
+            depthFirstOrder.get(b).add(count++);
+        }
+        return depthFirstOrder;
+    }
+
+    @Override
+    public String visualizeStore(S store) {
+        return store.visualize(this);
+    }
+
+    @Override
+    public String visualizeStoreFooter() {
+        return ")";
```

Comment:
Without the corresponding `visualizeStoreHeader` this is hard to keep track of. I would move this method to the two concrete classes.